lib/phoenix_live_view/async.ex

```diff
@@ -0,0 +1,209 @@
+defmodule Phoenix.LiveView.Async do
+  @moduledoc false
+
+  alias Phoenix.LiveView.{AsyncResult, Socket, Channel}
+
+  def start_async(%Socket{} = socket, name, func)
+      when is_atom(name) and is_function(func, 0) do
+    run_async_task(socket, name, func, :start)
+  end
+
+  def assign_async(%Socket{} = socket, key_or_keys, func)
+      when (is_atom(key_or_keys) or is_list(key_or_keys)) and
+             is_function(func, 0) do
+    keys = List.wrap(key_or_keys)
+
+    # verifies result inside task
+    wrapped_func = fn ->
+      case func.() do
+        {:ok, %{} = assigns} ->
+          if Enum.find(keys, &(not is_map_key(assigns, &1))) do
+            raise ArgumentError, """
+            expected assign_async to return map of assigns for all keys
+            in #{inspect(keys)}, but got: #{inspect(assigns)}
+            """
+          else
+            {:ok, assigns}
+          end
+
+        {:error, reason} ->
+          {:error, reason}
+
+        other ->
+          raise ArgumentError, """
+          expected assign_async to return {:ok, map} of
+          assigns for #{inspect(keys)} or {:error, reason}, got: #{inspect(other)}
+          """
+      end
+    end
+
+    new_assigns =
+      Enum.flat_map(keys, fn key ->
+        case socket.assigns do
+          %{^key => %AsyncResult{ok?: true} = _existing} -> []
+          %{} -> [{key, AsyncResult.new(key, keys)}]
+        end
+      end)
+
+    socket
+    |> Phoenix.Component.assign(new_assigns)
+    |> run_async_task(keys, wrapped_func, :assign)
+  end
+
+  defp run_async_task(%Socket{} = socket, keys, func, kind) do
+    if Phoenix.LiveView.connected?(socket) do
+      socket = cancel_existing(socket, keys)
+      lv_pid = self()
+      cid = cid(socket)
+      ref = make_ref()
+      {:ok, pid} = Task.start_link(fn -> do_async(lv_pid, cid, ref, keys, func, kind) end)
+      update_private_async(socket, &Map.put(&1, keys, {ref, pid, kind}))
+    else
+      socket
+    end
+  end
+
+  defp do_async(lv_pid, cid, ref, keys, func, async_kind) do
+    try do
+      result = func.()
+      Channel.report_async_result(lv_pid, async_kind, ref, cid, keys, {:ok, result})
+    catch
+      catch_kind, reason ->
+        Process.unlink(lv_pid)
+        caught_result = {:catch, catch_kind, reason, __STACKTRACE__}
```

Comment:
Nitpick: I would normalize the catch here, and not in the channel, so we send less information up.